Buy/Graph.CacheItem.swift

```diff
@@ -0,0 +1,132 @@
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+
+internal extension Graph {
+    
+    internal class CacheItem {
+        
+        let hash:      Hash
+        let data:      Data
+        let timestamp: Double
+        
+        // ----------------------------------
+        //  MARK: - Init -
+        //
+        init(hash: Hash, data: Data, timestamp: Double = Date().timeIntervalSince1970) {
+            self.hash      = hash
+            self.data      = data
+            self.timestamp = timestamp
+        }
+        
+        init?(at location: Location) {
+            guard let result = CacheItem.read(from: location) else {
+                return nil
+            }
+            
+            self.data      = result.data
+            self.hash      = result.meta["hash"]      as! String
+            self.timestamp = result.meta["timestamp"] as! Double
+        }
+        
+        // ----------------------------------
+        //  MARK: - IO -
+        //
+        private static func read(from location: Location) -> (data: Data, meta: [String : Any])? {
+            guard let data = try? Data(contentsOf: location.dataURL),
+                let meta = try? Data(contentsOf: location.metaURL),
+                let json = (try? JSONSerialization.jsonObject(with: meta, options: [])) as? [String : Any] else {
+                    
+                    return nil
+            }
+            
+            return (data, json)
+        }
+        
+        @discardableResult
+        func write(to location: Location) -> Bool {
+            let metaJson: [String : Any] = [
+                "hash"      : self.hash,
+                "timestamp" : self.timestamp
+            ]
+            
+            do {
+                let metaData = try JSONSerialization.data(withJSONObject: metaJson, options: [])
+                
+                try metaData.write(to: location.metaURL, options: .atomic)
+                try self.data.write(to: location.dataURL, options: .atomic)
+                
+            } catch {
+                print("Failed to flush CacheItem: \(error)")
+                return false
+            }
+            
+            return true
+        }
+    }
+}
+
+// ----------------------------------
+//  MARK: - Location -
+//
+extension Graph.CacheItem {
+    internal struct Location {
+        
+        let dataURL: URL
+        let metaURL: URL
+        
+        // ----------------------------------
+        //  MARK: - Init -
+        //
+        init(inParent url: URL, hash: Graph.Hash) {
+            let dataURL = url.appendingPathComponent(hash)
+            self.init(
+                dataURL: dataURL,
+                metaURL: dataURL.appendingPathExtension("meta")
+            )
+        }
+        
+        init(dataURL: URL, metaURL: URL) {
```

Comment:
That doesn't necessarily have to be enforced since those locations can be different. The `metaURL = dataURL + .meta` condition is only enforced by the other initialer that explicitly takes a `parentDirectory` as a parameter. Having said that, it doesn't hurt to make this `private`.